### PR TITLE
One-click GCP stack deployments

### PR DESCRIPTION
**Pull request by stefannica**
## Describe changes
Adding GCP to the list of cloud providers for which ZenML supports "one-click" full stack deployments - ZenML stacks with attached infrastructure provisioned through automated in-browser cloud-specific mechanisms, in this case [GCP Cloud Shell](https://cloud.google.com/shell/?gad_source=1&gclid=CjwKCAjw4ri0BhAvEiwA8oo6F9ZzkWiesxKoBDMPKQHSHaDjpC_qCC_1xeuNeU_QC45q9KWIA5sABxoC5k4QAvD_BwE&gclsrc=aw.ds) and [GCP Deployment Manager](https://cloud.google.com/deployment-manager/docs?gad_source=1&gclid=CjwKCAjw4ri0BhAvEiwA8oo6F2ainJeG1U7R9scHi9Lszf6mCtJp6_xNJJzNS5XxhNYTI-taAFplchoCEJIQAvD_BwE&gclsrc=aw.ds).

This allows users to easily deploy a full GCP ZenML stack with all the associated infrastructure and credentials:

* from the CLI with the `zenml stack deploy -p gcp` command
* from the dashboard, using the "Smart stack setup" workflows

The GCP ZenML stack includes:

* S3 bucket as artifact store
* Artifact Registry as container registry
* Vertex AI as orchestrator
* GCP Cloud Build as image builder

Example CLI output:

```
$ zenml stack deploy -p gcp --set --name gcp-stefan --location europe-west2

GCP ZenML Cloud Stack Deployment                                                                                                                                                                                   
================================

Provision and register a basic GCP ZenML stack authenticated and connected to all the necessary cloud infrastructure resources required to run pipelines in GCP.                                                   

Instructions                                                                                                                                                                                                       

You will be redirected to a GCP Cloud Shell console in your browser where you'll be asked to log into your GCP project and then create a Deployment Manager deployment to provision the necessary cloud resources  
for ZenML.                                                                                                                                                                                                         

NOTE: The Deployment Manager deployment will create the following new resources in your GCP project. Please ensure you have the necessary permissions and are aware of any potential costs:                        

 • A GCS bucket registered as a ZenML artifact store.                                                                                                                                                              
 • A Google Artifact Registry registered as a ZenML container registry.                                                                                                                                            
 • Vertex AI registered as a ZenML orchestrator.                                                                                                                                                                   
 • GCP Cloud Build registered as a ZenML image builder.                                                                                                                                                            
 • A GCP Service Account with the minimum necessary permissions to access the above resources.                                                                                                                     
 • An GCP Service Account access key used to give access to ZenML to connect to the above resources through a ZenML service connector.                                                                             

The Deployment Manager deployment will automatically create a GCP Service Account secret key and will share it with ZenML to give it permission to access the resources created by the stack. You can revoke these 
permissions at any time by deleting the Deployment Manager deployment in the GCP Cloud Console.                                                                                                                    

Estimated costs                                                                                                                                                                                                    

A small training job would cost around: $0.60                                                                                                                                                                      

These are rough estimates and actual costs may vary based on your usage and specific GCP pricing. Some services may be eligible for the GCP Free Tier. Use the GCP Pricing Calculator for a detailed estimate based
on your usage.                                                                                                                                                                                                     

⚠️ The Cloud Shell session will warn you that the ZenML GitHub repository is untrusted. We recommend that you review the contents of the repository and then check the Trust repo checkbox to proceed with the      
deployment, otherwise the Cloud Shell session will not be authenticated to access your GCP projects.                                                                                                               

💡 After the Deployment Manager deployment is complete, you can close the Cloud Shell session and return to the CLI to view details about the associated ZenML stack automatically registered with ZenML.          

Configuration                                                                                                                                                                                                      

You will be asked to provide the following configuration values during the deployment process:                                                                                                                     

 
### BEGIN CONFIGURATION ###
ZENML_STACK_NAME=my-stack
ZENML_STACK_REGION=europe-west2
ZENML_SERVER_URL=https://...-zenml.cloudinfra.zenml.io
ZENML_SERVER_API_TOKEN=....
### END CONFIGURATION ###


Proceed to continue with the deployment. You will be automatically redirected to GCP in your browser. [y/n]: y
If your browser did not open automatically, please open the following URL into your browser to deploy the stack to GCP: GCP Cloud Shell Console.                                                                   


Waiting for the deployment to complete and the stack to be registered. Press CTRL+C to abort...
```

### Implementation Details

This PR mainly uses the existing code structure already put in place with the introduction of AWS full stack deployments in the previous release, with the following minor modifications:

* the stack deployment API endpoint is modified to also return an optional configuration string in addition to the deployment URL. Users are responsible for copy-pasting this configuration from the CLI/dashboard into the cloud provider console in cases which don't allow transferring this information automatically through URL query params (GCP is one of those cases).
* the stack deployment API endpoint also returns the list of integrations required by the stack
* the GCP service connector had to be adjusted to allow credentials to be passed in base64-encoded format

The GCP stack is provisioned through a GCP Cloud Shell session with an included MarkDown tutorial and bash script that the user needs to update and execute manually to deploy a GCP Deployment Manager template. This complication was required due to the following factors:

* no way to trigger a GCP Deployment Manager template creation directly through a URL
* no way to include custom parameters in the GCP Cloud Shell URL


### Side-changes

* fix the "publish AWS stack deployment templates" github action
* introduce a similar github action step for the GCP related artifacts
* add local image builder to the full AWS ZenML stack registered by the AWS cloud formation template (for completion sake)
* include the AWS account name in the name of the S3 bucket provisioned by the AWS cloud formation template. This is done to avoid S3 namespace clashing (note: all S3 bucket names share the same global namespace regardless of AWS account, region or owner).
* include more instruction in the `zenml stack deploy` CLI after the full stack is deployed: how to activate the stack, which integrations to install
* include cost estimation in the `zenml stack deploy` CLI
* add support to allow GCP service connector credentials to be provided as base64 encoded JSON strings. This is necessary because this is the ONLY format supported by the GCP Deployment Manager service account key resource.
* increase the lifetime of the ZenML API token generated for cloud stack deployments from 1 hour to 6 hours, to give users more time to go through the stack deployment process.
* fix a minor bug in the zenml DB upgrade logic that caused DB upgrade failures in development
* validate that the stack component flavors mentioned in the "full stack registration" endpoint are valid

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [x] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Comment by stefannica** [comment]
@coderabbitai review

File: docs/book/how-to/stack-deployment/deploy-a-cloud-stack.md [hunk 1]
```diff
@@ -62,8 +62,7 @@ Once the configuration is finished, you will see a deployment page:
 
 ![Deploying the new stack](../../.gitbook/assets/deploy_stack_aws_2.png)
 
-During this process, you  will be redirected you to a Cloud Formation page 
-on AWS. 
+During this process, you will be redirected to a Cloud Formation page on AWS.
```
Comment (avishniakov) [inline docs/book/how-to/stack-deployment/deploy-a-cloud-stack.md:65]:
```suggestion
During this process, you will be redirected to a Cloud Formation page on AWS Console. 
```

**Review by avishniakov — APPROVED** [review]

**Comment by avishniakov** [comment]
There are some issue detected by the CI, like docstrings. Please, make sure to sort them out before merging.

File: src/zenml/stack_deployments/stack_deployment.py [hunk 2]
```diff
@@ -89,42 +105,104 @@
...
+        """
+        return StackDeploymentInfo(
+            provider=cls.provider,
+            description=cls.description(),
+            instructions=cls.instructions(),
+            post_deploy_instructions=cls.post_deploy_instructions(),
+            integrations=cls.integrations(),
+            permissions=cls.permissions(),
+            locations=cls.locations(),
+        )
```
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/stack_deployments/stack_deployment.py:123]: no errors should be thrown

File: src/zenml/zen_stores/rest_zen_store.py [hunk 3]
```diff
@@ -2859,40 +2860,31 @@ def get_stack_deployment_info(
...
-        body = self.get(f"{STACK_DEPLOYMENT}{URL}", params=params)
-
-        if not isinstance(body, list) or len(body) != 2:
-            raise ValueError(
-                "Bad response body received from the stack deployment URL "
-                "endpoint."
-            )
-        return body[0], body[1]
+        body = self.get(f"{STACK_DEPLOYMENT}{CONFIG}", params=params)
+        return StackDeploymentConfig.model_validate(body)
```
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/zen_stores/rest_zen_store.py:2887]: already handled in the inner methods

**Pull request merged**